src/adsbinarynode.jl

```diff
@@ -0,0 +1,381 @@
+# This file is part of the TaylorIntegration.jl package; MIT licensed
+
+using StaticArrays: SVector
+import AbstractTrees: HasNodeType, NodeType, ParentLinks, StoredParents, TreeIterator,
+                      children, nodetype, nodevalue, parent, printnode, PreOrderDFS
+import Base: IteratorEltype, HasEltype, in, eltype, split
+
+if !isdefined(Base, :isnothing)        # Julia 1.0 support
+    using AbstractTrees: isnothing
+end
+
+export ADSDomain
+
+"""
+    ADSDomain{N, T <: Real}
+
+`N` dimensional box used in Automatic Domain Splitting to represent the
+domain of `N` jet transport variables.
+
+# Fields
+
+- `lo::SVector{N, T}`: lower bounds.
+- `hi::SVector{N, T}`: upper bounds.
+"""
+struct ADSDomain{N, T <: Real}
+    lo::SVector{N, T}
+    hi::SVector{N, T}
+    # Inner constructor
+    function ADSDomain{N, T}(lo::SVector{N, T}, hi::SVector{N, T}) where {N, T <: Real}
+        @assert all(hi .> lo) "Each upper bound must be greater than the respective lower bound"
+        new{N, T}(lo, hi)
+    end
+end
+
+# Outer constructors
+ADSDomain(lo::SVector{N, T}, hi::SVector{N, T}) where {N, T <: Real} = ADSDomain{N, T}(lo, hi)
+
+function ADSDomain(x::Tuple{T, T}...) where {T <: Real}
+    N = length(x)
+    return ADSDomain{N, T}(
+        SVector{N, T}(x[i][1] for i in eachindex(x)),
+        SVector{N, T}(x[i][2] for i in eachindex(x))
+    )
+end
+
+# Print method for ADSDomain
+# Example:
+# [-1.0, 1.0]×[-1.0, 1.0]
+function show(io::IO, s::ADSDomain{N, T}) where {N, T <: Real}
+    x = Vector{String}(undef, N)
+    for i in eachindex(x)
+        x[i] = string(SVector{2, T}(s.lo[i], s.hi[i]))
+    end
+    print(io, join(x, "×"))
+end
+
+# Return the diameter of each direction in s
+diams(s::ADSDomain{N, T}) where {N, T <: Real} = s.hi - s.lo
+# Return the lower bounds in s
+infimums(s::ADSDomain{N, T}) where {N, T <: Real} = s.lo
+# Return the upper bounds in s
+supremums(s::ADSDomain{N, T}) where {N, T <: Real} = s.hi
+# Split s in half along direction i
+function split(s::ADSDomain{N, T}, i::Int) where {N, T <: Real}
+    @assert 1 <= i <= N
+    mid = (s.lo[i] + s.hi[i])/2
+    a = ADSDomain{N, T}(
+        s.lo,
+        SVector{N, T}(i == j ? mid : s.hi[j] for j in 1:N)
+    )
+    b = ADSDomain{N, T}(
+        SVector{N, T}(i == j ? mid : s.lo[j] for j in 1:N),
+        s.hi
+    )
```

Comment:
I don't see a way to avoid the two `fors`, given that `SVector` is immutable. However, this is not that bad, since the `SVector` is built upon a generator and this function is not critical to the performance.